File: 5_Moving_Horizon_Estimation/2nd_order_linear/Python/main_mhe.py
import tclab
import numpy as np
import time
from APMonitor.apm import *
import matplotlib.pyplot as plt

# Connect to Arduino
a = tclab.TCLab()

# Run time in minutes
run_time = 10.0

# Number of cycles (1 cycle per 2 seconds)
loops = int(30.0*run_time)

# Temperature (K)
T1 = np.ones(loops) * a.T1 # measured T
T1mhe = np.ones(loops) * a.T1 # measured T
Tsp1 = np.ones(loops) * a.T1 # measured T
Kp = np.ones(loops) * 0.3598
tau = np.ones(loops) * 47.73
zeta = np.ones(loops) * 1.56
TC_ss = np.ones(loops) * 23

# milli-volts input
Q1 = np.ones(loops) * 0.0
Q2 = np.ones(loops) * 0.0
Q1[6:] = 100.0
Q1[50:] = 20.0
Q1[100:] = 80.0
Q1[150:] = 10.0
Q1[200:] = 95.0
Q1[250:] = 0.0

# time
tm = np.zeros(loops)

# moving horizon estimation
from mhe import *
s = 'http://byu.apmonitor.com'
b = 'mhe'
mhe_init()

# Main Loop
start_time = time.time()
prev_time = start_time

try:
    for i in range(loops-1):
        # Sleep time
        sleep_max = 2.0 
        sleep = sleep_max - (time.time() - prev_time)
        if sleep>=0.01:
            time.sleep(sleep)
        else:
            time.sleep(0.01)

        # Record time and change in time
        t = time.time()
        tm[i] = t - start_time
        dt = t - prev_time
        prev_time = t
                    
        # Read temperature in degC
        T1[i] = a.T1

        # MHE - match model to measurements
        # run MHE every 2 seconds (see mhe.csv)
        params = mhe(T1[i],Q1[i-1])
        Kp[i] = params[0]
        tau[i] = params[1]
        zeta[i] = params[2]
        TC_ss[i] = params[3]
        T1mhe[i] = params[4]

        # Write output (0-100%)
        Q1[i] = min(100.0,max(0.0,Q1[i]))
        a.Q1(Q1[i])

        # Plot
        plt.clf()
        ax=plt.subplot(4,1,1)
        ax.grid()
        plt.plot(tm[0:i],T1[0:i],'ro',MarkerSize=3,label=r'$T_1 Measured$')
        plt.plot(tm[0:i],T1mhe[0:i],'bx',MarkerSize=3,label=r'$T_1 MHE$')
        plt.ylabel('Temperature (degC)')
        plt.legend(loc='best')

        ax=plt.subplot(4,1,2)
        ax.grid()
        plt.plot(tm[0:i],Q1[0:i],'r-',MarkerSize=3,label=r'$Q_1$')
        plt.plot(tm[0:i],Q2[0:i],'b:',MarkerSize=3,label=r'$Q_2$')
        plt.ylabel('Heaters')
        plt.legend(loc='best')

        ax=plt.subplot(4,1,3)
        ax.grid()
        plt.plot(tm[0:i],Kp[0:i],'ro',MarkerSize=3,label=r'$K_p$')
        plt.plot(tm[0:i],zeta[0:i],'bx',MarkerSize=3,label=r'$\zeta$')
        plt.ylabel('Parameters')
        plt.legend(loc='best')

        ax=plt.subplot(4,1,4)
        ax.grid()
        plt.plot(tm[0:i],tau[0:i],'k^',MarkerSize=3,label=r'$\tau$')
        plt.plot(tm[0:i],TC_ss[0:i],'gs',MarkerSize=3,label=r'$TC_{ss}$')
        plt.ylabel('Parameters')
        plt.legend(loc='best')

        plt.xlabel('Time (sec)')
        plt.draw()
        plt.pause(0.05)

        # Open Web Interface
        if i==5:
            apm_web(s,b)

# Allow user to end loop with Ctrl-C           
except KeyboardInterrupt:
    # Disconnect from Arduino
    a.Q1(0)
    a.Q2(0)
    print('Shutting down')
    a.close()
    
# Make sure serial connection still closes when there's an error
except:           
    # Disconnect from Arduino
    a.Q1(0)
    a.Q2(0)
    print('Error: Shutting down')
    a.close()
    raise

File: 5_Moving_Horizon_Estimation/2nd_order_linear/MATLAB/main_mhe.m
clear all; close all; clc

addpath('apm')

s = 'http://byu.apmonitor.com';
b = 'mhe';

% Connect to Arduino
tclab;

% Run time in minutes
run_time = 5.0;

% Number of cycles (1 cycle per 2 seconds)
loops = round(30*run_time);

% Temperature (degC)
T1 = ones(1,loops) * T1C(); % measured T
T1mhe = ones(1,loops) * T1C(); % measured T
Kp = ones(1,loops) * 0.3598;
tau = ones(1,loops) * 47.73;
zeta = ones(1,loops) * 1.5;
TC_ss = ones(1,loops) * 23;
time = zeros(1,loops);

% milli-volts input
Q1 = zeros(1,loops);
Q2 = zeros(1,loops);
Q1(6:end) = 100.0;
Q1(50:end) = 20.0;
Q1(100:end) = 80.0;
Q1(150:end) = 10.0;
Q1(200:end) = 95.0;
Q1(250:end) = 0.0;

% time
tm = zeros(1,loops);

% moving horizon estimation
mhe_init();

start_time = clock;
prev_time = start_time;

% dynamic plot (note: subplots needs to be declared here first)
figure(1)
subplot(3,1,1)
hold on, grid on
anexp1 = animatedline('LineStyle','-', 'Color', 'k', 'LineWidth', 2);
anpred1 = animatedline('LineStyle','--','Color', 'r','LineWidth', 2);
ylabel('Temperature \circC')
legend('T_1 Measured', 'T_1 Predicted', ...
    'Location', 'northwest')
title('Temperature Estimation')
subplot(3,1,2)
hold on, grid on
anQ1 = animatedline('LineStyle','-', 'Color', 'k', 'LineWidth', 2);
anQ2 = animatedline('LineStyle','--', 'Color', 'b', 'LineWidth', 2);
ylabel('Power Level Q (%)')
legend('Q_1', 'Q_2', 'Location', 'northwest')
subplot(3,1,3)
hold on, grid on
anK = animatedline('LineStyle','-', 'Color', 'r', 'LineWidth', 2);
antau = animatedline('LineStyle','--', 'Color', 'b', 'LineWidth', 2);
anzeta = animatedline('LineStyle','-', 'Color', 'g', 'LineWidth', 2);
ylabel('Parameters')
legend('K x 100', 'tau', 'zeta x 100', 'Location', 'northwest')
xlabel('Time (sec)')

for ii = 1:loops
    % adjust power level
    h1(Q1(ii));
    h2(Q2(ii));
    
    % Pause Sleep time
    pause_max = 2.0;
    pause_time = pause_max - etime(clock,prev_time);
    if pause_time >= 0.0
        pause(pause_time - 0.01)
    else
        pause(0.01)
    end
    
    % Record time and change in time
    t = clock;
    dt = etime(t,prev_time);
    if ii>=2
        time(ii) = time(ii-1) + dt;
    end
    prev_time = t;

    % read and record from temperature controller
    T1(ii) = T1C();
    T2(ii) = T2C();
    
    % non-linear energy balance
    jj = ii+1;
    params = mhe(T1(ii),Q1(ii));
    Kp(jj) = params(1);
    tau(jj) = params(2);
    zeta(jj) = params(3);
    T1mhe(jj) = params(5);        
        
    % plot
    addpoints(anexp1,time(ii),T1(ii))
    addpoints(anpred1,time(ii),T1mhe(ii))
    addpoints(anQ1,time(ii),Q1(ii))
    addpoints(anQ2,time(ii),Q2(ii))
    addpoints(anK,time(ii),100*Kp(ii))
    addpoints(antau,time(ii),tau(ii))
    addpoints(anzeta,time(ii),100*zeta(ii))
    drawnow    
    
    % open web-interface
    if ii==10
        apm_web(s,b);
    end
end

h1(0);
h2(0);
disp('Heaters off')
% turn off heater but keep LED on if T > 50
if (T1C() || T2C()) > 50
    led(1)
    disp(['Warning, heater temperature 1 =', num2str(T1C())])
    disp(['Warning, heater temperature 2 =', num2str(T2C())])
else
    led(0)
end

% save txt file with data
data = [time',Q1',Q2',T1',T2'];
csvwrite('data.txt',data);